usecases/vessel-detection/index.md

```diff
@@ -0,0 +1,211 @@
+
+Initial Implementation > Containerise > Reduce Dependencies > Xarray > OpenEO
+
+### Adaptive Thresholding in Xarray
+
+After investigating the code initially provided by Planetek, we arrived at this xarray implementation of adaptive thresholding.
+
+```python
+WINDOW_LAT_SIZE = 30
+WINDOW_LON_SIZE = 30
+THRESHOLD_FACT = 4
+out = xr.ones_like(data)
+rolling_mean = data.rolling(
+    longitude=WINDOW_LAT_SIZE, latitude=WINDOW_LON_SIZE, center=True
+).mean()
+thresholded_image = data > rolling_mean * THRESHOLD_FACT
+raster = out.where(thresholded_image == True, other=0)
+```
+
+We are going to port this into an OpenEO process graph.
+
+### 1. Authenticate
+
+
+```python
+import openeo
+
+backend = "openeo.cloud"
+conn = openeo.connect(backend)
+conn = conn.authenticate_oidc()
+```
+
+### 2. Pre-processing prep
+
+This process graph requires either a sea or land mask to remove the land from our AOI. A seamask has been prepared for this example and is available on github. It was derived using the [Water Bodies](https://land.copernicus.eu/content/corine-land-cover-nomenclature-guidelines/html/) classes from Corine Land Cover.
+
+
+```python
+import folium
+
+# We are going to use the following GeoJson to operate as a sea mask for out process graph.
+SEA_MASK = "https://raw.githubusercontent.com/SerRichard/sea_mask/main/sea-mask-4326.json"
+
+fig = folium.Figure(width=600, height=400)
+map = folium.Map(location=[44.465488, 12.602316], zoom_start=8)
+folium.GeoJson(SEA_MASK).add_to(map)
+fig.add_child(map)
+```
+![SeaMask to be used](./seamask.png)
+
+### 3. Defining a process graph to detect vessels in our AOI.
+
+The spatial_extent we have defined includes, but is not limited to the sea mask that we will be using. We will use both available polarizations from the SENTINEL1_GRD collection, and run this graph on a little over a weeks worth of data.
+
+
+```python
+spatial_extent  = {
+          "west": 12.194377989297493,
+          "east": 12.758093271633888,
+          "south": 44.24420099164355,
+          "north": 44.85455845353388
+        }
+
+temporal_extent = [
+          "2021-10-01",
+          "2021-10-09"
+        ]
+
+collection      = "SENTINEL1_GRD"
+bands           = ["VV","VH"]
+
+s1_datacube = conn.load_collection(
+    collection,
+    spatial_extent=spatial_extent,
+    bands=bands,
+    temporal_extent=temporal_extent
+)
+```
+
+Load the geometries as a vector cube.
+
+
+```python
+sea_mask = s1_datacube.process("load_vector_cube", {"URL": SEA_MASK})
+```
+
+Replace the values that lie outside of our polygon with NaN values.
+
+
+```python
+masked_data = s1_datacube.mask_polygon(sea_mask)
+```
+
+Apply a window to our datacube, and per pixel, update the pixel's value to be the mean of the window surrounding it.
+
+This would be the equivalent of the `data.rolling()` from the xarray example above.
+
+
+```python
+aggregated_window = masked_data.aggregate_spatial_window(size=[30, 30], reducer="mean")
```

Comment:
From reading the description of aggregate_spatial_window, it is not the equivalent of 'rolling'. It would actually compute the mean over non-overlapping windows of 30x30 pixels, and replace each window by a single pixel containing the mean.
I think you need `apply_kernel`with a kernel of 31x31 pixels with all values set to 1/(30x30).

Reply:
a 2D convolution operation (apply_kernel) is in fact a weighted average, so to compute the mean/average, just set all weights to the same value.

Reply:
Then my mistake, but can produce the same output in the executor using apply_kernel, will switch to using that process. Need to patch a bug in the apply_kernel for processes dask first.